### DISCUSSION TITLE
PDM callback still called after PDM.end()

### DISCUSSION THREAD

**Discussion by TomKong666** [earlephilhower/arduino-pico#1156] [General]
Hi,
thank you very much for this awesome project.

I ran into diffculties when trying to write to flash with LittleFS from Core1 while Microphone data is received by Core1 in a callback function registered by PDM.onReceive(onPDMdata). Core1 then randomly stops sending messages over the rp2040.fifo. In onPDMdata I'm just counting the received data, to see it's running. The difficulties vanish if I don't start the Microphone ( don't call PM.begin() ) at Programm start. Having the Microphone running again, then stopping it by PDM.end() before writing to Flash, PDM callback function is still triggered and Core1 randomly stops, observed as described.

Adding _dma_channel_set_irq0_enabled(_dmaChannel, false);_ at the first line of PDMClass::end() in PDM.cpp, at least symptomatically solves the problem for now.

Here my Questions:
1) Is PDM.end() intended to stop all operations related to the Microphone ?
2) If yes, would it additionally be necessary to unload the related pio program ?
3) May be off topic but a question though, what are the preconditions I have to met before writing to flash ?

Thanks for any advice welcome.

**Comment by earlephilhower** [comment]
I don't actually have a Nano Connect RP2040, so have never tried the PDM.  It was pulled in from the MBED core several months ago.  Looking at the code, I think maybe there have been significant updates upstream that could be pulled in.  For example, your `::end` problem.  There was a PR upstream that did just what you said (disable IRQ):
https://github.com/arduino/ArduinoCore-mbed/commit/1725a602f52d3e16237116f60b07d1549e305697

Would you like to do a PR pulling in that 1 change (or your version of it using the SDK vs NVIC)?  You might also want to see if you have the bandwidth to re-port the entire PDM library.   It seems there were lots of changes since we diverged!


As for (3), the conditions should all be done for you if you're using the included libs, but
1. Freeze the other core to a tight RAM code loop (the other core can't read code from flash while this core writes)
2. Ensure any IRQs on this core are fully in RAM (including, for example, any STL code!) since this core also can't read code from flash while writing

Other than that, I think there are no other limitations.  There are the usual caveats about multicore on Arduino, namely don't try and call the same library from different cores because Arduino doesn't generally support multicore (or even multithread) parallelism.
